newless.js

```diff
@@ -70,10 +74,17 @@
     else {
       var instantiator = function() {};
       return function construct(constructor, args, target) {
+        if (arguments.length === 3 && typeof target !== 'function')
+          throw new TypeError(''+target+' is not a constructor');
         instantiator.prototype = (target || constructor).prototype;
         var instance = new instantiator();
         var value = constructor.apply(instance, args);
-        return (typeof value === "object" && value) || instance;
+        if (typeof value === "object" && value) {
+          // we can do better if __proto__ is available (in some ES5 environments)
+          if (value.__proto__) value.__proto__ = (target || constructor).prototype;
+          return value;
+        }
```

Comment:
I’m not sure I understand what this is supposed to be doing.

First, if the constructor returns an object, that return value should be the result and its prototype should not be modified. For example:

```js
function SomeClass () {
  return {thisIsNot: 'A SomeClass instance'};
}

var directInstance = new SomeClass();
console.log(directInstance instanceof SomeClass); // false

var reflectInstance = Reflect.construct(SomeClass, []);
console.log(reflectInstance instanceof SomeClass); // false
```

So I thought “this just needs to go after line 80 and alter `instance` instead of `value`.” But I’m pretty sure setting `instantiator.prototype` already does everything that would do. Am I missing something?